docs/src/appendix/packaging.md

```diff
@@ -0,0 +1,501 @@
+of the package, as needed, see [Directory Layout](#directory-layout). However, the contents
+of the package can be directly extracted from the package file without extracting all of the
+contents. See [Binary Format](#binary-format) for details.
+
+## Top-Level Metadata
+
+* Schema Version
+  * The version of the package format itself, currently `1`.
+* Package Name (fully-qualified, e.g. `namespace/package`)
+* Package Version (using semver2 semantics)
+* Minimum required Miden VM version
+* Package Type (Executable vs Library)
+* Start function/entrypoint (optional for libraries)
+  * For program packages, specifies the entrypoint, i.e. `main`
+  * For library packages, if given, specifies a function that must be called to initialize the package before any of its exports may be called within a given context
+* Features (optional extras)
+  * Documentation
+  * Debug information
+* Dependencies (name, version, digest)
+
+## Type Descriptor
+
+The type descriptor file is the canonical, binary form of a WebAssembly Component Model [type definition](https://github.com/WebAssembly/component-model/blob/main/design/mvp/Explainer.md#type-definitions).
+It is derived from a [WebAssembly Interface Types](https://github.com/WebAssembly/component-model/blob/main/design/mvp/WIT.md)
+file, maintained by hand by the package author.
+
+This descriptor is used for two purposes:
+
+1. To generate language-specific bindings to the interface(s) described. WIT is an Interface Description Language (IDL),
+describing interfaces in a language-agnostic way, but defined in terms of a [Canonical ABI](https://github.com/WebAssembly/component-model/blob/main/design/mvp/CanonicalABI.md),
+allowing efficient cross-language interop. We follow the semantics of the Wasm canonical ABI, but layered on top of
+our canonical Miden ABI, which describes how the Wasm abstract machine is translated to the Miden abstract machine.
+
+2. To provide the compiler with the necessary type information needed to emit lifting/lowering code at call boundaries,
+as well as mapping imports (from the packages' dependencies) to their corresponding MAST roots, so that calls use those
+roots rather than the fully-qualified names. This ensures that the code that was compiled and published is always the
+exact code that gets executed, in perpetuity.
+
+Because the type descriptor is in binary form, to view it in the more human-friendly WIT format, you
+will need to use [wasm-tools](https://github.com/bytecodealliance/wasm-tools), like so:
+
+    wasm-tools component wit extracted-package/component.wasm
+
+The contents of `component.wasm` here is not an actual Wasm module, but simply the binary representation
+of the WIT package.
+
+## Directory Layout
+
+The canonical layout of a package when extracted as a directory, is as follows:
+
+    name@version/
+    |- manifest.toml
+    |- component.wasm
+    |- lib/
+       |- 0x00000000.mast
+       |- 0x00000001.mast
```

Comment:
Not exactly. Each `.mast` file corresponds to a MAST root, but that root might be referenced from other roots in the same package. The purpose of breaking them up this way is to:

1.) Facilitate installation in the object store, de-duplicated by MAST root. When "installing" a package, the contents of `lib` are more or less inserted directly into the object store, except the parts which are already present in the store. For example, two helper functions, whether in the same package/project or not, which happen to compile to the same MAST root will only appear once in the object store (and the package, if applicable).
2.) Allow loading just the parts of a MAST forest which are needed when the VM is loading code for a MAST root that is not in its procedure cache yet. The VM can directly load the MAST for that root from the object store this way. If the MAST references other MAST roots, those can either be loaded optimistically into the cache, or deferred until the code is actually being executed. The loader does not need to load/process the entire MAST forest of a package and then discard what is not needed.

In the binary form, it's all one big MAST forest, it's simply when extracted to disk that we break it up this way. Since a `MastForest` can represent both a single tree and a forest, this means the code for loading MAST, executing it, and then fetching the parts that aren't present in that `MastForest` is identical whether you loaded an entire package worth of code, or just code for a single procedure.

Reply:
One thing that is still not clear to me is what is the relationship of the `lib` directory and the format described in https://github.com/0xPolygonMiden/compiler/pull/132.

It seems like the serialized version of `lib` would correspond to `EncodedMastForest`. But how to we break up a single `EncodedMastForest` into multiple `.mast` files (each of each, I'm assuming, is also serialized as `EncodedMastForest`).